### DISCUSSION TITLE
Problem with connection to Supabase Realtime Database using WebSocket and Arduino ESP32

### DISCUSSION THREAD

**Discussion by tracemodule** [supabase/supabase#30762] [Questions]
I am trying to use an Arduino ESP32 module to connect to a Supabase Realtime Database using WebSocket. The goal is to listen to updates in the database and send them to the ESP32 module in real time.

I am using the ArduinoWebsockets library to handle the WebSocket connection and the ESPSupabase library to interact with the Supabase database.

What I Need Help With
WebSocket Authentication:
I'm having trouble establishing and maintaining a WebSocket connection to the Supabase Realtime API.
Handling Real-Time Events:
I want the ESP32 to correctly parse incoming messages and perform actions based on them.
What is a WebSocket?
A WebSocket is a communication protocol that allows two-way, real-time communication between a client (e.g., ESP32) and a server (e.g., Supabase Realtime). It is ideal for applications that require frequent updates, like monitoring real-time data changes in a database.

Code (Simplified for Security)
Below is a simplified version of my code for reference. Sensitive information such as API keys and URLs has been omitted.

```
#include <Arduino.h>
#include <WiFi.h>
#include <TinyGPS++.h>
#include <ESPSupabase.h>
#include <ArduinoWebsockets.h>

// Wi-Fi credentials
const char *ssid = "YourWiFiSSID";
const char *password = "YourWiFiPassword";

// Supabase credentials
const char *supabaseURL = "https://your-supabase-url.supabase.co";  // Replace with your Supabase URL
const char *supabaseKey = "YourSupabaseAnonKey";                   // Replace with your Supabase anon key

// WebSocket endpoint
const char *supabaseRealtimeURL = "wss://[project-ref].supabase.co/realtime/v1/websocket?apikey=[anon-token]&log_level=info&vsn=1.0.0";

websockets::WebsocketsClient websocketClient;
Supabase supabaseClient;

// GPS settings
TinyGPSPlus gps;
#define GPS_RX_PIN 34
#define GPS_TX_PIN 33
HardwareSerial gpsSerial(1);

// Function to connect to Wi-Fi
void connectToWiFi() {
  Serial.print("Connecting to Wi-Fi");
  WiFi.begin(ssid, password);
  while (WiFi.status() != WL_CONNECTED) {
    delay(500);
    Serial.print(".");
  }
  Serial.println("\nWi-Fi connected!");
}

// Function to initialize WebSocket
void initializeWebSocket() {
  websocketClient.onMessage([](websockets::WebsocketsMessage message) {
    Serial.println("Realtime Update:");
    Serial.println(message.data());
  });

  Serial.println("Connecting to Supabase Realtime...");
  if (websocketClient.connect(supabaseRealtimeURL)) {
    Serial.println("Connected to Supabase Realtime!");
  } else {
    Serial.println("Failed to connect to Supabase Realtime!");
  }
}

void setup() {
  Serial.begin(115200);
  gpsSerial.begin(9600, SERIAL_8N1, GPS_RX_PIN, GPS_TX_PIN);

  // Connect to Wi-Fi
  connectToWiFi();

  // Initialize Supabase
  supabaseClient.begin(String(supabaseURL), supabaseKey);

  // Connect to WebSocket
  initializeWebSocket();
}

void loop() {
  // Keep WebSocket connection alive
  websocketClient.poll();
  delay(1000);
}
```

**Comment by Sylvesta** [comment]
Has any progress been made on this? I have tried multiple ESP32's and can't get a WebSocket connection either.
